applications/CoSimulationApplication/custom_utilities/feti_dynamic_coupling_utilities.cpp

```diff
@@ -0,0 +1,696 @@
+                            array_1d<double, 3>& r_nodal_quantity = rNode.FastGetSolutionStepValue(rVariable);
+                            for (size_t dof_dim = 0; dof_dim < dim; ++dof_dim)
+                            {
+                                r_nodal_quantity[dof_dim] += rCorrection[equation_id + dof_dim];
+                            }
+                        }
+                    }
+                }
+            );
+        }
+
+        KRATOS_CATCH("")
+    }
+
+    void FetiDynamicCouplingUtilities::WriteLagrangeMultiplierResults(const Vector& rLagrange)
+    {
+        KRATOS_TRY
+
+        const SizeType dim = mpOriginDomain->ElementsBegin()->GetGeometry().WorkingSpaceDimension();
+
+        ModelPart& r_interface = mrDestinationInterfaceModelPart;
+        auto interface_nodes = r_interface.NodesArray();
+        for (size_t i = 0; i < interface_nodes.size(); ++i)
+        {
+            IndexType interface_id = interface_nodes[i]->GetValue(INTERFACE_EQUATION_ID);
+
+            array_1d<double, 3>& lagrange = interface_nodes[i]->FastGetSolutionStepValue(VECTOR_LAGRANGE_MULTIPLIER);
+            lagrange.clear();
+            for (size_t dof = 0; dof < dim; dof++)
+            {
+                lagrange[dof] = -1.0*rLagrange[interface_id * dim + dof];
+            }
+        }
+
+        KRATOS_CATCH("")
+    }
+
+    void FetiDynamicCouplingUtilities::GetInterfaceQuantity(
+        ModelPart& rInterface, const Variable<array_1d<double, 3>>& rVariable,
+        Vector& rContainer, const SizeType nDOFs)
+    {
+        KRATOS_TRY
+
+        if (rContainer.size() != rInterface.NumberOfNodes() * nDOFs)
+            rContainer.resize(rInterface.NumberOfNodes() * nDOFs);
+        rContainer.clear();
+
+        KRATOS_ERROR_IF_NOT(rInterface.NodesArray()[0]->Has(INTERFACE_EQUATION_ID))
+            << "FetiDynamicCouplingUtilities::GetInterfaceQuantity | The interface nodes do not have an interface equation ID.\n"
+            << "This is created by the mapper.\n";
+
+        // Fill up container
+        block_for_each(rInterface.Nodes(), [&](Node<3>& rNode)
+            {
+                IndexType interface_id = rNode.GetValue(INTERFACE_EQUATION_ID);
+                array_1d<double, 3>& r_quantity = rNode.FastGetSolutionStepValue(rVariable);
+                for (size_t dof = 0; dof < nDOFs; dof++)  rContainer[nDOFs * interface_id + dof] = r_quantity[dof];
+            }
+        );
+
+        KRATOS_CATCH("")
+    }
+
+    void FetiDynamicCouplingUtilities::GetInterfaceQuantity(
+        ModelPart& rInterface, const Variable<double>& rVariable,
+        Vector& rContainer, const SizeType nDOFs)
+    {
+        KRATOS_TRY
+
+        if (rContainer.size() != rInterface.NumberOfNodes())
+            rContainer.resize(rInterface.NumberOfNodes());
+        else rContainer.clear();
+
+        KRATOS_ERROR_IF_NOT(rInterface.NodesArray()[0]->Has(INTERFACE_EQUATION_ID))
```

Comment:
I am surprised that this works, this shouldn't work if Node with Id 0 doesn't exist